File: CodingDojo/games/bomberman/src/main/cpp/bomberman/Direction.cpp
/*-
 * #%L
 * Codenjoy - it's a dojo-like platform from developers to developers.
 * %%
 * Copyright (C) 2018 Codenjoy
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */
#include "Direction.h"


Direction::Direction(int value) {
	auto i = Directions.find(valueOf(value));
	if (i == Directions.end()) {
		throw std::invalid_argument("Direction::Direction(int value): No direction for " + value);
	}
	dir = *i;
}

Direction::Direction(String name) {
	auto i = Directions.find(name);
	if (i == Directions.end()) {
		throw std::invalid_argument("Direction::Direction(String name): No direction for such name ");
	}
	dir = *i;
}

bool Direction::operator==(const Direction& d) const {
	return d.dir == dir;
}

bool Direction::operator!=(const Direction& d) const {
	return !operator==(d);
}

bool Direction::isNull() const {
	return dir.first == LL("NULL");
}

String Direction::toString() const {
	StringStream ss;
	ss <<  (isNull() ? LL("") : dir.first);
	return ss.str();
}

String Direction::valueOf(int value) {
	for (auto i : Directions) {
		if (std::get<0>(i.second) == value) return i.first;
	}
	throw std::invalid_argument("Direction::valueOf(int value): No direction for " + value);
}

int Direction::changeX(int x) const {
	return x + std::get<1>(dir.second);
}

int Direction::changeY(int y) const {
	return y - std::get<2>(dir.second);
}

Direction Direction::inverted() const {
	if (dir.first == LL("UP")) return Direction(LL("DOWN"));
	if (dir.first == LL("DOWN")) return Direction(LL("UP"));
	if (dir.first == LL("LEFT")) return Direction(LL("RIGHT"));
	if (dir.first == LL("RIGHT")) return Direction(LL("LEFT"));
	return Direction(LL("STOP"));
}


DirectionMap Direction::initialise() {
	DirectionMap mapOfDirections;
	mapOfDirections[LL("NULL")] = DirectionVector(-1, 0, 0);

	// direction of Bomberman
	mapOfDirections[LL("UP")] = DirectionVector(2, 0, -1);
	mapOfDirections[LL("DOWN")] = DirectionVector(3, 0, 1);
	mapOfDirections[LL("LEFT")] = DirectionVector(0, -1, 0);
	mapOfDirections[LL("RIGHT")] = DirectionVector(1, 1, 0);

	// drop a bomb
	mapOfDirections[LL("ACT")] = DirectionVector(4, 0, 0);

	// stop the Bomberman
	mapOfDirections[LL("STOP")] = DirectionVector(5, 0, 0);
	return mapOfDirections;
}

const DirectionMap Direction::Directions = Direction::initialise();

File: CodingDojo/games/minesweeper/src/main/javascript/Direction.js
/*-
 * #%L
 * Codenjoy - it's a dojo-like platform from developers to developers.
 * %%
 * Copyright (C) 2018 - 2020 Codenjoy
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */
var D = function (index, dx, dy, name, isAction) {

    var withAct = function(){
        if (!! isAction) return D(index, dx, dy, name, !!isAction);
        return D(index, 0, 0, name + ',act', true);
    };

    var changeX = function (x) {
        return x + dx;
    };

    var changeY = function (y) {
        return y - dy;
    };

    var inverted = function () {
        switch (this) {
            case Direction.UP: return Direction.DOWN;
            case Direction.DOWN: return Direction.UP;
            case Direction.LEFT: return Direction.RIGHT;
            case Direction.RIGHT: return Direction.LEFT;
            default: return Direction.STOP;
        }
    };

    var toString = function () {
        return name;
    };

    return {
        changeX: changeX,

        changeY: changeY,

        inverted: inverted,

        toString: toString,

        getIndex: function () {
            return index;
        },

        withAct: withAct,

        isAction: isAction

    };
};

var Direction = {
    UP: D(2, 0, -1, 'up', false),                 // you can move
    DOWN: D(3, 0, 1, 'down', false),
    LEFT: D(0, -1, 0, 'left', false),
    RIGHT: D(1, 1, 0, 'right', false),
    ACT: D(4, 0, 0, 'act', true),                // drop bomb
    STOP: D(5, 0, 0, '', true)                   // stay
};

Direction.values = function () {
    return [Direction.UP, Direction.DOWN, Direction.LEFT, Direction.RIGHT, Direction.ACT, Direction.STOP];
};

Direction.valueOf = function (index) {
    var directions = Direction.values();
    for (var i in directions) {
        var direction = directions[i];
        if (direction.getIndex() == index) {
            return direction;
        }
    }
    return Direction.STOP;
};


if (module) module.exports = Direction;
